File: web/src/api/interview.ts
/**
 * Interview Review API Client
 * Handles all API calls related to interview review feature
 * 
 * 注意：client.ts 的响应拦截器已经解包了 axios response，
 * 返回的是 API 响应体 { code, data, msg }，所以这里只需要访问 .data 即可获取实际数据
 */

import apiClient from './client';
import type { ApiResponse } from '@/types/global';
import type {
  InterviewReview,
  InterviewReviewListResponse,
  CreateReviewRequest,
  UpdateReviewMetadataRequest,
  ListReviewsParams,
} from '@/types/interview';

/**
 * Create a new interview review record
 * 在TOS上传完成后调用，创建pending状态的记录
 * @param data - Review data including tos_file_key and audio_filename
 * @returns Created review record
 */
export const createReview = async (data: CreateReviewRequest): Promise<InterviewReview> => {
  const response = await apiClient.post<ApiResponse<InterviewReview>>('/api/interview/reviews', data);
  // response 是 { code, data, msg }，response.data 是 InterviewReview
  return (response as unknown as ApiResponse<InterviewReview>).data;
};

/**
 * Get a single interview review by ID
 * @param id - Review ID
 * @returns Interview review record
 */
export const getReview = async (id: number): Promise<InterviewReview> => {
  const response = await apiClient.get<ApiResponse<InterviewReview>>(`/api/interview/reviews/${id}`);
  return (response as unknown as ApiResponse<InterviewReview>).data;
};

/**
 * List interview reviews with pagination
 * @param params - Query parameters (page, page_size, status)
 * @returns Paginated list of reviews
 */
export const listReviews = async (params?: ListReviewsParams): Promise<InterviewReviewListResponse> => {
  const response = await apiClient.get<ApiResponse<InterviewReviewListResponse>>('/api/interview/reviews', {
    params: {
      page: params?.page || 1,
      page_size: params?.page_size || 10,
      ...(params?.status && { status: params.status }),
    },
  });
  return (response as unknown as ApiResponse<InterviewReviewListResponse>).data;
};

/**
 * Start ASR (speech recognition) for a review
 * Uses the stored tos_file_key to generate a temporary URL and submit an ASR task
 * @param id - Review ID
 * @returns Updated review record with transcribing status
 */
export const startASR = async (id: number): Promise<InterviewReview> => {
  const response = await apiClient.post<ApiResponse<InterviewReview>>(`/api/interview/reviews/${id}/start-asr`);
  return (response as unknown as ApiResponse<InterviewReview>).data;
};

/**
 * Retry ASR (speech recognition) for a review
 * Used after ASR failure, generates a new temporary URL and submits a new ASR task
 * @param id - Review ID
 * @returns Updated review record with transcribing status
 */
export const retryASR = async (id: number): Promise<InterviewReview> => {
  const response = await apiClient.post<ApiResponse<InterviewReview>>(`/api/interview/reviews/${id}/retry-asr`);
  return (response as unknown as ApiResponse<InterviewReview>).data;
};

/**
 * Sync ASR result from asr_tasks table to interview_review metadata
 * Backend directly fetches from ASR table and updates review metadata
 * @param id - Review ID
 * @returns Updated review record with asr_result in metadata
 */
export const syncASRResult = async (id: number): Promise<InterviewReview> => {
  const response = await apiClient.post<ApiResponse<InterviewReview>>(`/api/interview/reviews/${id}/sync-asr`);
  return (response as unknown as ApiResponse<InterviewReview>).data;
};

/**
 * Trigger AI analysis for a review
 * Requires ASR to be completed (asr_result must exist in metadata)
 * @param id - Review ID
 * @returns Updated review record with analyzing status
 */
export const triggerAnalysis = async (id: number): Promise<InterviewReview> => {
  const response = await apiClient.post<ApiResponse<InterviewReview>>(`/api/interview/reviews/${id}/analyze`);
  return (response as unknown as ApiResponse<InterviewReview>).data;
};

/**
 * Update interview review metadata (job position, company, etc.)
 * @param id - Review ID
 * @param data - Metadata fields to update
 * @returns Updated review record
 */
export const updateReviewMetadata = async (
  id: number,
  data: UpdateReviewMetadataRequest
): Promise<InterviewReview> => {
  const response = await apiClient.patch<ApiResponse<InterviewReview>>(`/api/interview/reviews/${id}`, data);
  return (response as unknown as ApiResponse<InterviewReview>).data;
};

/**
 * Export all interview API methods
 */
export const interviewAPI = {
  createReview,
  getReview,
  listReviews,
  startASR,
  retryASR,
  syncASRResult,
  triggerAnalysis,
  updateReviewMetadata,
};

export default interviewAPI;

File: server/api/interview/interview.go
package interview

import (
	"server/service/interview"
	"server/utils"
	"strconv"

	"github.com/gin-gonic/gin"
)

// CreateReviewRequest 创建面试复盘记录请求
// 基于TOS文件信息创建，ASR任务将在后续步骤中触发
type CreateReviewRequest struct {
	TosFileKey    string `json:"tos_file_key" binding:"required"`   // TOS文件key
	AudioFilename string `json:"audio_filename" binding:"required"` // 音频文件名
}

// CreateReview 创建面试复盘记录
// 在TOS上传完成后立即调用，创建pending状态的记录
func CreateReview(c *gin.Context) {
	// 获取用户ID
	userID := c.GetString("userID")
	if userID == "" {
		utils.FailWithMessage("无法识别用户", c)
		return
	}

	var req CreateReviewRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.FailWithMessage("请求参数错误: "+err.Error(), c)
		return
	}

	// 创建记录
	review, err := interview.InterviewService.CreateInterviewReview(userID, req.TosFileKey, req.AudioFilename)
	if err != nil {
		utils.FailWithMessage(err.Error(), c)
		return
	}

	utils.OkWithData(review, c)
}

// GetReview 获取面试复盘记录详情
func GetReview(c *gin.Context) {
	// 获取用户ID
	userID := c.GetString("userID")
	if userID == "" {
		utils.FailWithMessage("无法识别用户", c)
		return
	}

	// 获取记录ID
	reviewIDStr := c.Param("id")
	reviewID, err := strconv.ParseInt(reviewIDStr, 10, 64)
	if err != nil {
		utils.FailWithMessage("记录ID格式错误", c)
		return
	}

	// 获取记录
	review, err := interview.InterviewService.GetInterviewReview(reviewID, userID)
	if err != nil {
		utils.FailWithMessage(err.Error(), c)
		return
	}

	utils.OkWithData(review, c)
}

// ListReviews 获取用户的面试复盘记录列表
func ListReviews(c *gin.Context) {
	// 获取用户ID
	userID := c.GetString("userID")
	if userID == "" {
		utils.FailWithMessage("无法识别用户", c)
		return
	}

	// 解析分页参数
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "10"))

	// 查询列表
	response, err := interview.InterviewService.ListInterviewReviews(userID, page, pageSize)
	if err != nil {
		utils.FailWithMessage(err.Error(), c)
		return
	}

	utils.OkWithData(response, c)
}

// UpdateReviewMetadataRequest 更新面试复盘元数据请求
type UpdateReviewMetadataRequest struct {
	Metadata map[string]interface{} `json:"metadata" binding:"required"` // 要更新的元数据字段
}

// UpdateReviewMetadata 更新面试复盘记录元数据（如岗位、公司等）
func UpdateReviewMetadata(c *gin.Context) {
	// 获取用户ID
	userID := c.GetString("userID")
	if userID == "" {
		utils.FailWithMessage("无法识别用户", c)
		return
	}

	// 获取记录ID
	reviewIDStr := c.Param("id")
	reviewID, err := strconv.ParseInt(reviewIDStr, 10, 64)
	if err != nil {
		utils.FailWithMessage("记录ID格式错误", c)
		return
	}

	var req UpdateReviewMetadataRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.FailWithMessage("请求参数错误: "+err.Error(), c)
		return
	}

	// 更新元数据
	review, err := interview.InterviewService.UpdateReviewMetadata(reviewID, userID, req.Metadata)
	if err != nil {
		utils.FailWithMessage(err.Error(), c)
		return
	}

	utils.OkWithData(review, c)
}

// TriggerAnalysis 触发面试分析工作流
func TriggerAnalysis(c *gin.Context) {
	// 获取用户ID
	userID := c.GetString("userID")
	if userID == "" {
		utils.FailWithMessage("无法识别用户", c)
		return
	}

	// 获取记录ID
	reviewIDStr := c.Param("id")
	reviewID, err := strconv.ParseInt(reviewIDStr, 10, 64)
	if err != nil {
		utils.FailWithMessage("记录ID格式错误", c)
		return
	}

	// 触发分析
	review, err := interview.InterviewService.TriggerAnalysis(reviewID, userID)
	if err != nil {
		utils.FailWithMessage(err.Error(), c)
		return
	}

	utils.OkWithData(review, c)
}

// StartASR 启动ASR语音识别任务
// 使用存储的tos_file_key生成临时URL并提交ASR任务
func StartASR(c *gin.Context) {
	// 获取用户ID
	userID := c.GetString("userID")
	if userID == "" {
		utils.FailWithMessage("无法识别用户", c)
		return
	}

	// 获取记录ID
	reviewIDStr := c.Param("id")
	reviewID, err := strconv.ParseInt(reviewIDStr, 10, 64)
	if err != nil {
		utils.FailWithMessage("记录ID格式错误", c)
		return
	}

	// 启动ASR
	review, err := interview.InterviewService.StartASR(reviewID, userID)
	if err != nil {
		utils.FailWithMessage(err.Error(), c)
		return
	}

	utils.OkWithData(review, c)
}

// RetryASR 重试ASR语音识别任务
// 在ASR失败后使用存储的tos_file_key生成新URL并重新提交任务
func RetryASR(c *gin.Context) {
	// 获取用户ID
	userID := c.GetString("userID")
	if userID == "" {
		utils.FailWithMessage("无法识别用户", c)
		return
	}

	// 获取记录ID
	reviewIDStr := c.Param("id")
	reviewID, err := strconv.ParseInt(reviewIDStr, 10, 64)
	if err != nil {
		utils.FailWithMessage("记录ID格式错误", c)
		return
	}

	// 重试ASR
	review, err := interview.InterviewService.RetryASR(reviewID, userID)
	if err != nil {
		utils.FailWithMessage(err.Error(), c)
		return
	}

	utils.OkWithData(review, c)
}

// SyncASRResult 从ASR表同步识别结果
// 直接从asr_tasks表获取数据并更新到interview_review的metadata中
func SyncASRResult(c *gin.Context) {
	// 获取用户ID
	userID := c.GetString("userID")
	if userID == "" {
		utils.FailWithMessage("无法识别用户", c)
		return
	}

	// 获取记录ID
	reviewIDStr := c.Param("id")
	reviewID, err := strconv.ParseInt(reviewIDStr, 10, 64)
	if err != nil {
		utils.FailWithMessage("记录ID格式错误", c)
		return
	}

	// 同步ASR结果
	review, err := interview.InterviewService.SyncASRResult(reviewID, userID)
	if err != nil {
		utils.FailWithMessage(err.Error(), c)
		return
	}

	utils.OkWithData(review, c)
}
